CRM/Volunteer/BAO/Project.php

```diff
@@ -237,59 +237,166 @@ static function isActive($entityId, $entityTable) {
-          'contact_id' => array("IN" => (array) $contactIds),
-          'relationship_type_id' => $relType,
-        ));
-        foreach ($api['values'] as $data) {
-          $projectIds[] = $data['project_id'];
-        }
+    $query = CRM_Utils_SQL_Select::from('`civicrm_volunteer_project` vp')
+      ->select('DISTINCT vp.*');
 
-        // if none of the passed contacts are related to any projects, then
-        // there's no need to continue
-        if (empty($projectIds)) {
-          return $result;
-        }
+    if (!empty($params['project_contacts'])) {
+      $contactJoin = self::buildContactJoin($params['project_contacts']);
+      if ($contactJoin) {
+        $query->join('vpc', $contactJoin);
       }
-      unset($params['project_contacts']);
     }
 
+    if (!empty($params['proximity'])) {
+      $query->join('loc', 'INNER JOIN `civicrm_loc_block` loc ON loc.id = vp.loc_block_id')
+        ->join('civicrm_address', 'INNER JOIN `civicrm_address` ON civicrm_address.id = loc.address_id')
+        ->where(self::buildProximityWhere($params['proximity']));
+    }
+
+    // This step is here to support both naming conventions for specifying params
+    // (e.g., volunteer_project_id and id) while normalizing how we access them
+    // (e.g., $project->id)
     $project = new CRM_Volunteer_BAO_Project();
     $project->copyValues($params);
 
-    if (!empty($projectIds)) {
-      $valuesSql = implode(', ', array_unique($projectIds));
-      $project->whereAdd(" id IN ({$valuesSql}) ");
+    foreach ($project->fields() as $field) {
+      $fieldName = $field['name'];
+
+      if (!empty($project->$fieldName)) {
+        $query->where('!column = @value', array(
+          'column' => $fieldName,
+          'value' => $project->$fieldName,
+        ));
+      }
     }
 
-    $project->find();
+    $dao = self::executeQuery($query->toSQL());
+    while ($dao->fetch()) {
+      $fetchedProject = new CRM_Volunteer_BAO_Project();
+      $fetchedProject->copyValues(clone $dao);
+      $result[(int) $dao->id] = $fetchedProject;
+    }
+    $dao->free();
+
+    return $result;
+  }
+
+  /**
+   * Helper method to filter Projects by related contact.
+   *
+   * Conditionally invoked by CRM_Volunteer_BAO_Project::retrieve().
+   *
+   * @param array $projectContacts
+   *   @see CRM_Volunteer_BAO_Project::create() for details on this parameter
+   * @return mixed
+   *   Boolean FALSE if no projects have the specified contact relationships;
+   *   String SQL fragment otherwise
+   */
+  private static function buildContactJoin(array $projectContacts) {
+    $result = FALSE;
+    $onClauses = array();
+
+    $relTypes = CRM_Core_OptionGroup::values(
+      CRM_Volunteer_BAO_ProjectContact::RELATIONSHIP_OPTION_GROUP,
+      TRUE, FALSE, FALSE, NULL, 'name');
 
-    while ($project->fetch()) {
-      $result[(int) $project->id] = clone $project;
+    foreach ($projectContacts as $relType => $contactIds) {
```

Comment:
There is not a limit, but the relationship types are Project Manager, Project Owner, and Project Beneficiary -- each can take multiple contacts, but I don't imagine that happening too often, and not with large numbers of contacts when it does.